### PR TITLE
June 13 Homework [Daniel Mourad]

### PR DESCRIPTION
**Pull request by mouradyan**
Please review my homework

### CODE REVIEW

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+securityGroupId=
+sshKeyName="ec2-key"
+sshKey="ec2-key.pem"
+instanceId=
+instancePublicIp=
+
+# creates vpc with 172.22.0.0/16 cidr block and sends vpc id to ResourceIds file
+function createVPC () {
+	echo "Creating VPC with 172.22.0.0/16 CIDR Block..."
+	aws ec2 create-vpc \
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:19]: assign aws command output directly to a variable(query must return exact VPCid without any further information)

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+		--cidr-block 172.22.0.0/16 \
+		--query Vpc.VpcId \
+		--output text > $ResourceIds && \
+	echo "VPC is created"
+	vpcId=$(cat $ResourceIds | head -1)
+}
+
+# creates subnet with 172.22.22.0/24 cidr block and sends subnet id to ResourceIds file
+function createSubnet () {
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:28]: write a better query to get exact Subnet ID
Dont create the additional file, use variable directly

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+		--vpc-id $vpcId \
+		--cidr-block 172.22.22.0/24 \
+		--query Subnet.SubnetId \
+		--output text >> $ResourceIds && \
+	echo "Subnet is created"
+	subnetId=$(cat $ResourceIds | head -2 | tail +2)
+}
+
+# creates internet gateway and sends the id to ResourceIds file
+function createInternetGateway () {
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:40]: same as for subnets

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	echo "Creating Internet Gateway..."
+	aws ec2 create-internet-gateway \
+		--query InternetGateway.InternetGatewayId \
+		--output text >> $ResourceIds && \
+	echo "Internet Gateway created"
+	internetGatewayId=$(cat $ResourceIds | head -3 | tail +3)
+}
+
+# attaches internet gateway to the vpc
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:49]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+# creates vpc with 172.22.0.0/16 cidr block and sends vpc id to ResourceIds file
+function createVPC () {
+	echo "Creating VPC with 172.22.0.0/16 CIDR Block..."
+	aws ec2 create-vpc \
+		--cidr-block 172.22.0.0/16 \
+		--query Vpc.VpcId \
+		--output text > $ResourceIds && \
+	echo "VPC is created"
+	vpcId=$(cat $ResourceIds | head -1)
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:25]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+function createSubnet () {
+	echo "Creating Subnet with 172.22.22.0/24 CIDR Block..."
+	aws ec2 create-subnet \
+		--vpc-id $vpcId \
+		--cidr-block 172.22.22.0/24 \
+		--query Subnet.SubnetId \
+		--output text >> $ResourceIds && \
+	echo "Subnet is created"
+	subnetId=$(cat $ResourceIds | head -2 | tail +2)
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:37]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+# creates internet gateway and sends the id to ResourceIds file
+function createInternetGateway () {
+	echo "Creating Internet Gateway..."
+	aws ec2 create-internet-gateway \
+		--query InternetGateway.InternetGatewayId \
+		--output text >> $ResourceIds && \
+	echo "Internet Gateway created"
+	internetGatewayId=$(cat $ResourceIds | head -3 | tail +3)
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:47]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+# attaches internet gateway to the vpc
+function attachInternetGatewayToVpc () {
+	echo "Attaching Internet Gateway to VPC..."
+	aws ec2 attach-internet-gateway \
+		--vpc-id $vpcId \
+		--internet-gateway-id $internetGatewayId \
+		--output text > /dev/null && \
+	echo "Internet Gateway is successfully attached to VPC"
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:57]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+# creates route table and sends the id to ResourceIds file
+function createRouteTable () {
+	echo "Creating Route Table..."
+	aws ec2 create-route-table \
+		--vpc-id $vpcId \
+		--query RouteTable.RouteTableId \
+		--output text >> $ResourceIds && \
+	echo "Route Table is created"
+	routeTableId=$(cat $ResourceIds | head -4 | tail +4)
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:68]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+# creates route to anywhere
+function createRoute () {
+	echo "Creating Route to 0.0.0.0/0..."
+	aws ec2 create-route \
+		--route-table-id $routeTableId \
+		--destination-cidr-block 0.0.0.0/0 \
+		--gateway-id $internetGatewayId \
+		--output text > /dev/null && \
+	echo "Route is created"
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:79]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+# associates route table with subnet
+function associateRouteTable () {
+	echo "Associating Route Table with Subnet"
+	aws ec2 associate-route-table \
+		--subnet-id $subnetId \
+		--route-table-id $routeTableId \
+		--output text > /dev/null && \
+	echo "Route Table is associated"
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:89]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+	echo "Creating Security Group for SSH and HTTP access..."
+	aws ec2 create-security-group \
+		--group-name SSH-HTTP-Access \
+		--description "Security group for SSH and HTTP access" \
+		--vpc-id $vpcId \
+		--query GroupId \
+		--output text >> $ResourceIds && \
+	echo "Security Group is created"
+	securityGroupId=$(cat $ResourceIds | head -5 | tail +5)
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:102]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+		--output text > /dev/null && \
+	echo "SSH access is authorized"
+	aws ec2 authorize-security-group-ingress \
+		--group-id $securityGroupId \
+		--protocol tcp \
+		--port 80 \
+		--cidr 0.0.0.0/0 \
+		--output text > /dev/null && \
+	echo "HTTP access is authorized"
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:121]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+		--key-name $sshKeyName \
+		--security-group-ids $securityGroupId \
+		--subnet-id $subnetId \
+		--associate-public-ip-address | grep InstanceId | cut -d '"' -f 4 >> $ResourceIds && \
+	echo "Instance is successfully created"
+	instanceId=$(cat $ResourceIds | head -6 | tail +6) 
+	aws ec2 describe-instances \
+        --instance-id $instanceId | grep "PublicIpAddress" | cut -d '"' -f 4 >> $ResourceIds
+	instancePublicIp=$(cat $ResourceIds | head -7 | tail +7)
+}
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:149]: fail scenario not handled

File: Homework-june-13/Daniel_Mourad/create-ec2.sh [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+	fi
+elif [[ $1 = "--help" ]]
+then
+	echo "--create ->  creates VPC, Subnet, Internet Gateway, Route Table, Security Group and EC2 Instance"
+	echo "--purge -> deletes all the created resources"
+	echo "--show-resources -> shows all the available resources"
+else
+	echo "You need to specify an option to continue"
+	echo "See --help for more information about options"
+fi
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:222]: add cleanup for any fail scenario

Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/create-ec2.sh:222]: script is not only creating ec2 instance, so change the scripts name

File: Homework-june-13/Daniel_Mourad/aws-ec2-automation.sh [hunk 16]
```diff
@@ -0,0 +1,414 @@
+#!/bin/bash
+
+set -e
```
Comment (Gev544) [inline Homework-june-13/Daniel_Mourad/aws-ec2-automation.sh:3]: will conflict with your own error handling

Comment (mouradyan) [reply Homework-june-13/Daniel_Mourad/aws-ec2-automation.sh:3]: The set +e used when needed so set -e will not conflict with error handling

**Review by Gev544 — APPROVED** [review]

**Pull request merged**